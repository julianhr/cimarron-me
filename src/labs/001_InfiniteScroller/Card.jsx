import React from 'react'
import PropTypes from 'prop-types'
import styled from '@emotion/styled'

import ImageLoader from '~/components/library/ImageLoader'


const Root = styled.article`
  width: 100%;
  border-radius: 10px;
  margin-bottom: 10px;
  padding: 20px 25px 10px;
  min-height: 140px;
`

const H4 = styled.h4`
  color: ${props => props.theme.colors.text};
  padding-bottom: 10px;
`

const P = styled.p`
  padding-top: 5px;
`

const imgStyle = {
  display: 'none',
  '@media (min-width: 520px)': {
    display: 'inherit',
    float: 'left',
    width: 150,
    height: 100,
    marginRight: 30,
    marginBottom: 10,
  }
}

function Card({ title, imgUrl, description, forwardedRef, position }) {
  const pElDescription = description.map((desc, i) =>
    <P key={i}>{desc}</P>
  )

  return (
    <Root
      ref={forwardedRef}
      css={{
        background: forwardedRef ? '#f0d5ea' : 'white',
        border: forwardedRef ? '1px solid #a07f99' : null,
      }}
    >
      <ImageLoader
        maxWidth={150}
        maxHeight={100}
        imgSrc={imgUrl}
        styles={{ root: imgStyle }}
      />
      <H4>{`${position}. ${title}`}</H4>
      {pElDescription}
    </Root>
  )
}

Card.propTypes = {
  description: PropTypes.array.isRequired,
  forwardedRef: PropTypes.object,
  imgUrl: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  position: PropTypes.number.isRequired,
}

export default Card
