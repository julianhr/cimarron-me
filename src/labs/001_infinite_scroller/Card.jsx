import React from 'react'
import PropTypes from 'prop-types'
import styled from '@emotion/styled'


const Root = styled.article`
  width: 100%;
  border-radius: 10px;
  margin-bottom: 10px;
  padding: 20px 25px 10px;
  min-height: 140px;
`

const H4 = styled.h4`
  color: ${props => props.theme.colors.text};
  padding-bottom: 10px;
`

const P = styled.p`
  padding-top: 5px;
`

const Img = styled.img`
  float: left;
  width: 180px;
  height: 100px;
  padding-right: 30px;
  padding-bottom: 4px;

  ${props => props.theme.queries.upTo('xs')} {
    display: none;
  }
`

function Card({ title, imgUrl, description, forwardedRef, position }) {
  const pElDescription = description.map((desc, i) =>
    <P key={i}>{desc}</P>
  )

  return (
    <Root
      ref={forwardedRef}
      css={{
        background: forwardedRef ? '#f0d5ea' : 'white',
        border: forwardedRef ? '1px solid #a07f99' : null,
      }}
    >
      <Img src={imgUrl} />
      <H4>{`${position}. ${title}`}</H4>
      {pElDescription}
    </Root>
  )
}

Card.propTypes = {
  description: PropTypes.array.isRequired,
  forwardedRef: PropTypes.object,
  imgUrl: PropTypes.string,
  title: PropTypes.string.isRequired,
  position: PropTypes.number,
}

export default Card
