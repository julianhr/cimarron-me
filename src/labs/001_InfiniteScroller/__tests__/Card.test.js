import React from 'react'
import { create } from 'react-test-renderer'
import emotionSerializer from 'jest-emotion'

import Card from '../Card'
import MockInfiniteScrollerApp from '../__tests__/__mocks__/MockInfiniteScrollerApp'


const renderApp = props => create(
  <MockInfiniteScrollerApp>
    <Card {...props} />
  </MockInfiniteScrollerApp>
)

describe('Card', () => {
  const testProps = {
    description: ['sentence 1', 'sentence 2'],
    forwardedRef: React.createRef(),
    imgUrl: 'http://test.com',
    title: 'test title',
    position: 1,
  }

  describe('props', () => {
    it('description is required', () => {
      const props = {...testProps}
      delete props.description
      expect(() => renderApp(props)).toThrow()
    })

    it('title is required', () => {
      const props = {...testProps}
      delete props.title
      expect(() => renderApp(props)).toThrow()
    })

    it('imgUrl is required', () => {
      const props = {...testProps}
      delete props.imgUrl
      expect(() => renderApp(props)).toThrow()
    })

    it('position is required', () => {
      const props = {...testProps}
      delete props.position
      expect(() => renderApp(props)).toThrow()
    })

    it('forwardedRef is not required', () => {
      const props = {...testProps}
      delete props.forwardedRef
      expect(() => renderApp(props)).not.toThrow()
    })
  })

  describe('snapshots', () => {
    beforeEach(() => {
      expect.addSnapshotSerializer(emotionSerializer)
    })

    it('all props', () => {
      expect(renderApp(testProps)).toMatchSnapshot()
    })

    it('required props', () => {
      const props = {...testProps}
      delete props.forwardedRef
      expect(renderApp(props)).toMatchSnapshot()
    })
  })

  it('connects forwardedRef to root element', () => {
    const props = {...testProps}
    props.forwardedRef = React.createRef()
    const wrapper = renderApp(props)
    const element = wrapper.root.findAllByType('article')
    expect(element).toHaveLength(1)
    expect(element[0].instance).toBe(props.forwardedRef.current)
  })
})
