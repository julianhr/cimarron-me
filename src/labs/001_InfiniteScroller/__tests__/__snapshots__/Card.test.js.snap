// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card snapshots all props 1`] = `
.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #FFF;
  display: none;
  width: 150px;
  height: 100px;
}

@media (min-width:520px) {
  .emotion-1 {
    display: inherit;
    float: left;
    width: 150px;
    height: 100px;
    margin-right: 30px;
    margin-bottom: 10px;
  }
}

.emotion-0 {
  -webkit-transition: opacity 80ms linear;
  transition: opacity 80ms linear;
  will-change: opacity;
  opacity: 0;
}

.emotion-3 {
  color: #222222;
  padding-bottom: 10px;
}

.emotion-5 {
  padding-top: 5px;
}

.emotion-9 {
  width: 100%;
  border-radius: 10px;
  margin-bottom: 10px;
  padding: 20px 25px 10px;
  min-height: 140px;
  background: #f0d5ea;
  border: 1px solid #a07f99;
}

<article
  className="emotion-9 emotion-10"
>
  <figure
    className="emotion-1 emotion-2"
  >
    <img
      className="emotion-0"
      onLoad={[Function]}
      src="http://test.com"
    />
  </figure>
  <h4
    className="emotion-3 emotion-4"
  >
    1. test title
  </h4>
  <p
    className="emotion-5 emotion-6"
  >
    sentence 1
  </p>
  <p
    className="emotion-5 emotion-6"
  >
    sentence 2
  </p>
</article>
`;

exports[`Card snapshots required props 1`] = `
.emotion-9 {
  width: 100%;
  border-radius: 10px;
  margin-bottom: 10px;
  padding: 20px 25px 10px;
  min-height: 140px;
  background: white;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #FFF;
  display: none;
  width: 150px;
  height: 100px;
}

@media (min-width:520px) {
  .emotion-1 {
    display: inherit;
    float: left;
    width: 150px;
    height: 100px;
    margin-right: 30px;
    margin-bottom: 10px;
  }
}

.emotion-0 {
  -webkit-transition: opacity 80ms linear;
  transition: opacity 80ms linear;
  will-change: opacity;
  opacity: 0;
}

.emotion-3 {
  color: #222222;
  padding-bottom: 10px;
}

.emotion-5 {
  padding-top: 5px;
}

<article
  className="emotion-9 emotion-10"
>
  <figure
    className="emotion-1 emotion-2"
  >
    <img
      className="emotion-0"
      onLoad={[Function]}
      src="http://test.com"
    />
  </figure>
  <h4
    className="emotion-3 emotion-4"
  >
    1. test title
  </h4>
  <p
    className="emotion-5 emotion-6"
  >
    sentence 1
  </p>
  <p
    className="emotion-5 emotion-6"
  >
    sentence 2
  </p>
</article>
`;
